### DISCUSSION TITLE
Expand Legacy Update to Windows 9x OSes

### DISCUSSION THREAD

**Discussion by tariqmudallal** [LegacyUpdate/LegacyUpdate#9] [Ideas]
First of all, I want to give props to you for being able to _even_ replicate the old Windows Update website and get it to work with the Automatic Update functionality in Windows 2000, XP, Vista & 7!

Now, I would like to make a feature request for the following: I would like to see you be able to bring Legacy Update to the Windows 9x OSes (95, 98 & Me). I think this would make for a great addition because even these websites had some form of ActiveX Control that would deal with downloading and installing updates from the web browser. Currently, there is no way to get _all_ the updates for the Win9x OSes, instead forcing users to manually scout for them all via _a lot of_ Google searching, or through Unofficial Service Packs. I think that Legacy Update for Win9x would make for a great addition since, for instance, if someone wants to keep the vanilla experience while being up-to-date, then I believe this would be the ideal solution!

**Comment by Blood-PawWerewolf** [comment]
Same here. Win 9x/Me support would definitely be appreciated.

**Comment by atauenis** [comment]
There are list of all (?) updates for Windows 98: https://web.archive.org/web/20060707123624/http://www.microsoft.com:80/windows98/downloads/corporate.asp . But this is only a update catalog, and not related to Automatic Updates.

Implementing automatic update support on Win9x is difficultier than Windows Update v6, but is not impossible. There was few automatic update services on Win9x:

1. Windows Update v4. Used by MS CUN Utility v4.0 in 2002-2011. Sadly, but it's protocol haven't dumped/sniffed when it was live. But some files are still available from MS servers.
   - An incomplete log: https://gist.github.com/atauenis/dd8574df7181b23b33e56906897f7a8e
   - The most important `https://v4.windowsupdate.microsoft.com/*manifest.asp` files are lost.
   - Haven't checked, but it may be similar to used by older versions of MS SUS? A cascade of WU>Win2012R2/WSUS6.3>Win2003/WSUS7.0>WinXP still worked in 2021, may be WSUS7.0>SUS3.0 also will work? :)
2. Windows Update v3. Used by MS CUN Tool before v4.0. Remains non-preserved.
   - There was something for Win95+IE5.5, which also installed some kind of MS ActiveX control. But I can't remember, it finds need updates, or simply was a catalog. AFAIR, the site design was completely different from one on Win98/ME.
3. Internet Explorer Automatic Update. Set of CDF files, downloaded by MSIE 4.0 from MS servers, Probably, CDFs have this syntax: https://learn.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa768160(v=vs.85) . 
4. MS IE ActiveX searcher. Used by MSIE 3.0 (and may be 4/5 too) to find unknown ActiveX controls installers. Well web-archived, and may be rebuilt.
5. Windows Media Player Update. Used by WMP 5.2 (beta of 6.0), 6.x-11. It was something XML-based, but I can't find any examples of the files.
6. Office Update. Seems that was not automatical, but there was some kind of ActiveX control used by it to find & install updates for Office 2000, XP (2002). Since Office 2003, updates were moved to MS Update v6/WSUS (and Office XP updates too).

**Reply by kirb** [reply]
Thanks for the research!

I've been playing around WU v3 in the past few weeks (https://twitter.com/hbkirb/status/1592484473320640515) and have made progress as far as getting the ActiveX control to do everything up to loading the catalog, which for some reason it refuses to do. Haven't figured out why yet.

I was able to create a modified ident.cab which pointed everything at a test server on my local network (this is why that "do you trust this" unsigned file warning appeared), and it successfully issued these requests:

```
test.legacyupdate.net 100.78.31.100 - - [15/Nov/2022:22:11:10 +1030] "GET /windowsupdate/v3/v31content/ident.cab?cdbfd4-fffadd3f-fffb02d7-0a100c2820118 HTTP/1.0" 200 1167 0.000s "-" "Windows Update"
test.legacyupdate.net 100.78.31.100 - - [15/Nov/2022:22:11:14 +1030] "GET /windowsupdate/v3/v31content/inventory.plt?cdbfd4-fffadd3f-fffb02d7-0a100c2820118 HTTP/1.0" 200 513 0.000s "-" "Windows Update"
test.legacyupdate.net 100.78.31.100 - - [15/Nov/2022:22:11:14 +1030] "GET /windowsupdate/v3/v31content/bitmask.plt?cdbfd4-fffadd3f-fffb02d7-0a100c2820118 HTTP/1.0" 200 425 0.000s "-" "Windows Update"
test.legacyupdate.net 100.78.31.100 - - [15/Nov/2022:22:11:14 +1030] "GET /windowsupdate/v3/v31content/FileVer.INI?cdbfd4-fffadd3f-fffb02d7-0a100c2820118 HTTP/1.0" 200 602 0.000s "-" "Windows Update"
test.legacyupdate.net 100.78.31.100 - - [15/Nov/2022:22:11:14 +1030] "GET /windowsupdate/v3/v31content/OSDET.W98?cdbfd4-fffadd3f-fffb02d7-0a100c2820118 HTTP/1.0" 200 3311 0.000s "-" "Windows Update"
```

But when invoking the catalog download function, it doesn't make any request (not even to Microsoft, confirmed with my proxy), and then jumps to an error page because the control reports that 0 items are in the catalog it never loaded (I commented out the error page redirect in that demo video).

I'd imagine the contents of these files are something to do with it, but they're a bit of a pain to reverse:

* inventory.plt is a cab containing some binary format I can't identify, there's no ascii text in there at all, but it does seem to follow a pattern.
* bitmask.plt is another unidentifiable binary format, very small file.
* FileVer.ini seems to identify the latest version of inseng.dll (delivered via e.g. inseng.w98, a cab containing the dll), but of course we don't see it downloading this file yet
* OSDET.W98 is another dll in a cab ("Windows Update Platform Detection" 5.3.357.0).

As for the actual update payloads, I think I should have a comprehensive bunch of them thanks to me combining all of the separate update site revisions Wayback Machine has archived. It helps that much of Wayback's old archives come from ISP cache backups, and I'm sure ISPs would have found it extremely important to be caching these big exes their customers are downloading because their computer is asking them to do so. But I won't really know what I have till I can actually browse the catalog properly through a working WU v3 interface.

I'd definitely be interested in looking into the other update mechanisms, especially the Office pre-XP one. (Might just roll that into WU v3 since editing the catalog is pretty easy.) WU v4 looks more promising in terms of getting the updater itself going (v3's implementation isn't amazing, and is flawed from a security perspective), but I'm convinced updates prior to Win2k are going to be mostly lost as I explain in the [FAQ](https://legacyupdate.net/faq), and will require rebuilding the WU backend and its database from scratch.

**Reply by kirb** [reply]
Just took a better look at your logs and first off - wow! Totally missed that it's from 2008. Crowdsourcing old WU logs might be a good idea, no idea how many people out there still have a working Win9x PC running an install from 10+ years ago though I guess.

What I was interested by was that the cab and exe update urls mentioned are still live. That's extremely good news. We just need to build a catalog. Googling "ie501dom.exe" brought up the KB tied to that update, so I guess an archive of the KBs would be a pretty decent starting point. "com_microsoft.CUN_40_98_recupd_5576" brought up [this](https://finet.sihvonen.narkive.com/6ZeBVhtO/samia-ei-nay-ei-kuulu#post12) incredible find! Looks like some kind of log file, no download links but welp, it's a start. Gives me more stuff I can google.

Do you still have access to the PC this is from? Can you zip and upload C:\\Program Files\\WindowsUpdate? (Feel free to reach me on Discord if you don't want to post it publicly, @kirb#1809, send a friend request first.)

**Reply by atauenis** [reply]
A full backup of all WU-related stuff:
https://disk.yandex.ru/d/jG_nO3Mz1mfRaw - 98
https://disk.yandex.ru/d/sOtQ-ZMPWuHoPA - ME (almost empty)
It was virtual machines I've used in 2008 to test unknown software before installing in main OS. I have been installed some unofficial service pack on 98 before opening WU, so probably this is why almost no updates were found by CUN/WU Agent.

In WinME (too from 2008) there are even no WindowsUpdate.log. I don't know why, but as I remember, I've just opened update.microsoft.com, accepted certificate, then closed it. But file dates are slightly newer than WinME RTM date (WUAUBoot.exe from 2001, ident.txt from 2003), so may be result of some WU work.

**Reply by sydia1103** [reply]
Hi there Kirb, I just wanted to let you know that a group of individuals have partially gotten Windows Update v5, v4, and v3 working. I know that LegacyUpdate runs v6, but updates for 98 and older would be helpful. But for right now, we have reversed engineered v4's getmanifest.asp, so it successfully checks for updates on 2000 RTM and later, however, results.asp is not working at the moment since that page was never archived. v3 works on 95 and later but only updates until Windows 2000. There is a dedicated server for this, if you want to join, you are welcome to, but you don't have to. Below are some images of what we have accomplished. ![image](https://user-images.githubusercontent.com/69560065/208309393-772bbb01-f774-4e78-a4bf-4ae86bb1c557.png)
![image](https://user-images.githubusercontent.com/69560065/208309401-144e5705-e771-499a-be93-14de7df512d1.png)

**Reply by kirb** [reply]
@sydia1103 That is extremely cool, I’m very interested. Feel free to send me a friend request @kirb#1809 and DM an invite.

**Reply by Athlonxp32** [reply]
i would say that`s extremley nice Works after so many years.. Today i found Kirb`s Site and i`m very happy

**Reply by kristibektashi** [reply]
Here's something interesting: Windows Media Player Update remnants still exist in WMP 12 (Windows 7 - 11) and it tries to connect to it when you launch `C:\Program Files (x86)\Windows Media Player\setup_wm.exe` (might be a different exe) but of course it shows an error

**Comment by Shaknunic** [comment]
Hey, I still have the real old Windows 98 and some updates from the original Security CD that Microsoft released in 2004... I wanted to help you out with this project if you are up for it. Not quite sure what you need, but I want to offer help... I tried to add you on Discord, but you are refusing invites atm so...

My XMPP: shaknunic@jabber.hot-chilli.net

**Reply by kirb** [reply]
Sorry about that, try again now.

**Comment by atauenis** [comment]
Found that MS Word 97 have some auto update feature (unlike I've thinked that there was no anything in it). In some cases (mostly when opening HTML files in it), Word attempts to download from `http://www.microsoft.com/isapi/msoffice/word/autoupdate.idc?Prd=word&Sbp=ia&Ar=AutoUpdate&Pver=8.0.0.5622&o1=3.0.3424.1&o2=0&o3=0x419&OS=WinX86.` This URL is configurable via registry ([Q163299](https://ftp.zx.net.nz/pub/archive/ftp.microsoft.com/MISC/KB/en-us/163/299.HTM)).

So, probably, if a `autoupdate.idc` cached output will be found, it will be theoretically possible to add Office 97 support. :)

**Comment by 7dog123** [comment]
They got the [Windows update restored.](http://windowsupdaterestored.com/)  project up and running.

**Comment by kristibektashi** [comment]
Like the commenter above me said, there is a site called https://windowsupdaterestored.com/ that has old versions of Windows Update archived. Perhaps @kirb could work with them and integrate it into Legacy Update so that it can be used on older Windows versions (and maybe merge the catalogs too so that all updates are available from a single website)

**Comment by kristibektashi** [comment]
Btw if Windows 9x is supported perhaps NT 4 could also be supported
